syntax = "proto3";

package robocup.humanoid;

import "google/protobuf/timestamp.proto";

/// A column vector of two floats
message fvec2 {
    float x = 1;
    float y = 2;
}

/// A column vector of three floats
message fvec3 {
    float x = 1;
    float y = 2;
    float z = 3;
}

/// A matrix of three vectors
/// Specified as column vectors
message fmat3 {
    fvec3 x = 1;
    fvec3 y = 2;
    fvec3 z = 3;
}

/// The detected team of the robot
enum Team {
    UNKNOWN_TEAM = 0;
    BLUE         = 1;
    RED          = 2;
}

message Robot {
    /// ID of the robot, 0 if not known
    uint32 player_id = 1;

    /// The position of the robot on the field according to the following convention
    /// x meters along the field with 0 at the centre of the field and positive towards opponent goals
    /// y meters across the field with 0 at the centre of the field and positive to the left
    /// θ orientation of the robot (anti-clockwise from the x axis from above the field)
    fvec3 position = 2;

    /// The covariance measure of the robots [x, y, θ] values
    /// If this is unavailable leave this unset
    fmat3 covariance = 3;

    /// Robot team, if known
    Team team = 4;
}

message Ball {
    /// The position of the ball on the field according to the following convention
    /// x meters along the field with 0 at the centre of the field and positive towards opponent goals
    /// y meters across the field with 0 at the centre of the field and positive to the left
    /// z meters above the field with 0 at the surface of the field and positive up
    fvec3 position = 1;

    /// The velocity of the ball
    /// x m/s along the field with 0 at the centre of the field and positive towards opponent goals
    /// y m/s across the field with 0 at the centre of the field and positive to the left
    /// z m/s above the field with 0 at the surface of the field and positive up
    /// Set to 0 if not available
    fvec3 velocity = 2;

    /// The covariance measure of the balls [x, y, z] values
    /// If this is unavailable leave this unset
    fmat3 covariance = 3;
}

/// The current playing state of the robot
enum State {
    UNKNOWN_STATE = 0;
    UNPENALISED   = 1;
    PENALISED     = 2;
}

/// The role of the robot
enum Role {
    ROLE_UNDEFINED = 0;
    ROLE_IDLING=1;
    ROLE_OTHER=2;
    ROLE_STRIKER=3;
    ROLE_SUPPORTER=4;
    ROLE_DEFENDER=5;
    ROLE_GOALIE=6;
}

/// The current action of the robot
enum Action {
    ACTION_UNDEFINED=0;
    ACTION_POSITIONING=1;
    ACTION_GOING_TO_BALL=2;
    ACTION_TRYING_TO_SCORE=3;
    ACTION_WAITING=4;
    ACTION_KICKING=5;
    ACTION_SEARCHING=6;
    ACTION_LOCALIZING=7;
}

/// The offensive strategy of the robot
enum OffensiveSide {
    SIDE_UNDEFINED = 0;
    SIDE_LEFT = 1;
    SIDE_MIDDLE = 2;
    SIDE_RIGHT = 3;
}

message Message {
    /// Timestamp of message creation
    google.protobuf.Timestamp timestamp = 1;

    /// The robots current state
    State state = 2;

    /// Position, orientation, and covariance of the player on the field
    Robot current_pose = 3;

    /// The current walk speed of the robot in it's local [x, y, θ] coordinates
    /// x and y in m/s and θ in rad/s
    /// positive x being forwards, positive y being strafing to the left, and positive θ being anti-clockwise
    fvec3 walk_command = 4;

    /// Position and orientation of the players target on the field specified
    Robot target_pose = 5;

    /// Position that the robot is aiming to kick the ball to
    /// If no kick is planned set to [0, 0]
    /// Vector has origin on the ground between the robots feet
    fvec2 kick_target = 6;

    /// Position, velocity, and covariance of the ball on the field
    Ball ball = 7;

    /// Position, orientation, and covariance of detected robots on the field
    repeated Robot others = 8;

    /// IDs 1-100 are reserved for official use

    /****************************
     * EXTENSIONS GO BELOW HERE *
     ****************************/

    /// Maximum walking speed of the robot
    float max_walking_speed = 101;

    /// Time the robot needs to reach the ball;
    float time_to_ball = 104;

    /// The role of the robot
    Role role = 105;

    /// The current action of the robot
    Action action = 106;

    /// The offensive strategy of the robot
    OffensiveSide offensive_side = 107;

    /// Confidence values of the obstacles
    repeated float other_robot_confidence = 108;
}
